components/sendgrid/sendgrid.app.js

```diff
@@ -108,5 +114,212 @@ module.exports = {
+          },
+        })
+      );
+    },
+    async addOrUpdateContacts(params) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          method: "PUT",
+          path: `/marketing/contacts`,
+          data: params,
+        })
+      );
+    },
+    async createContactList(name) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          method: "POST",
+          path: `/marketing/lists`,
+          data: {
+            name,
+          },
+        })
+      );
+    },
+    async deleteBlocks(deleteAll, emails) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          method: "DELETE",
+          path: `/suppression/blocks`,
+          data: {
+            delete_all: deleteAll,
+            emails,
+          },
+        })
+      );
+    },
+    async deleteBounces(deleteAll, emails) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          method: "DELETE",
+          path: `/suppression/bounces`,
+          data: {
+            delete_all: deleteAll,
+            emails,
+          },
+        })
+      );
+    },
+    async deleteContacts(deleteAllContacts, ids) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          method: "DELETE",
+          path: `/marketing/contacts`,
+          params: {
+            delete_all_contacts: deleteAllContacts,
+            ids,
+          },
+        })
+      );
+    },
+    async deleteGlobalSupression(email) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          method: "DELETE",
+          path: `/asm/suppressions/global/${email}`,
+        })
+      );
+    },
+    async deleteList(id, deleteContacts) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          method: "DELETE",
+          path: `/marketing/lists/${id}`,
+          params: {
+            delete_contacts: deleteContacts,
+          },
+        })
+      );
+    },
+    async getBlock(email) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          path: `/suppression/blocks/${email}`,
+        })
+      );
+    },
+    async getGlobalSupression(email) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          path: `/asm/suppressions/global/${email}`,
+        })
+      );
+    },
+    async getAllBounces(startTime, endTime) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          path: `/suppression/bounces`,
+          params: {
+            start_time: startTime,
+            end_time: endTime,
+          },
+        })
+      );
+    },
+    async getAllContactLists(pageSize, pageToken) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          path: `/marketing/lists`,
+          params: {
+            page_size: pageSize,
+            page_token: pageToken,
+          },
+        })
+      );
+    },
```

Comment:
As mentioned above, `pageSize` and `pageToken` are API details and should not be exposed by this method.
Instead, we should accept a `maxItems` argument to use as an upper bound on the collection of retrieved records.

```suggestion
    async *getAllContactLists(maxItems) {
      let url = `${this._apiUrl()}/marketing/lists`;

      while (url && maxItems > 0) {    
        const params = {
          page_size: Math.min(maxItems, 1000),
        };
        const requestConfig = {
          ...this._makeRequestConfig(),
          params,
        };
        const { data } = this._withRetries(
          () => axios.get(url, requestConfig),
        );

        const contactLists = data.result.slice(0, maxItems);
        for (const contactList of contactLists) {
          yield contactList
        }

        maxItems -= contactLists.length;
        url = data._metadata.next;
      }
    },
```

Reply:
Understood. Will be using similar approach on other actions related to an API endpoint having pageSize/pageToken..